File: C-Sharp/Data Structures and Algorithms/StacksAndQueues/StacksAndQueues/StacksAndQueues/Queue.cs
using System;
using System.Text;

namespace StacksAndQueues
{
    public class Queue<Type>
    {
        public Node<Type> Front { get; set; }
        public Node<Type> Rear { get; set; }

        public Queue()
        {
            Rear = Front;
        }

        /// <summary>
        /// Addes a Node to the Rear of the queue
        /// </summary>
        /// <param name="value">The value of the added Node</param>
        public void Enqueue(Type value)
        {
            Node<Type> temp = new Node<Type>(value);
            if (Rear == null)
            {
                Front = temp;
                Rear = temp;
            }
            else
            {
                Rear.Next = temp;
                Rear = temp;
            }
        }

        /// <summary>
        /// Removes a Node from the Front of the queue
        /// </summary>
        /// <returns>Value of the removed Node</returns>
        public Type Dequeue()
        {
            if (Front == null)
                throw new NullReferenceException();
            else if (Front.Next == null)
            {
                Type temp = Front.Value;
                Front = null;
                Rear = Front;
                return temp;
            }
            else
            {
                Type temp = Front.Value;
                Front = Front.Next;
                return temp;
            }
        }

        /// <summary>
        /// Looks at the value of the Front Node
        /// </summary>
        /// <returns>Value of the Front Node</returns>
        public Type Peek()
        {
            return Front.Value;
        }

        /// <summary>
        /// Checks whether or not a queue is empty
        /// </summary>
        /// <returns>True if empty, False if not empty</returns>
        public bool IsEmpty()
        {
            if (Front == null) return true;
            else return false;
        }
    }
}

File: Java/src/main/java/challenges/utilities/Queue.java
package challenges.utilities;

public class Queue<T> {
    private Node<T> front = null;
    private Node<T> rear = null;

    public boolean isEmpty() { return front == null; }

    public T peek() { return isEmpty() ? null : front.value; }

    public void enqueue(T value) {
        Node<T> node = new Node(value);
        if (isEmpty())
            front = node;
        else
            rear.next = node;
        rear = node;
    }

    public T dequeue() {
        if (isEmpty()) { return null; }
        Node<T> node = front;
        front = front.next;
        return node.value;
    }
}
